File: Dart/lib/data_structures/lists/entities/singly_linked_list.dart
import 'package:dart/util/interfaces/linked_list.dart';

import 'package:dart/data_structures/models/node.dart';

import 'package:dart/util/functions/get_node.dart';

class SinglyLinkedList<T> implements LinkedList<T> {
  Node<T>? _head;
  Node<T>? _tail;
  int _counter = 0;

  T? get head => _head?.el;

  T? get tail => _tail?.el;

  int get length => _counter;

  int push({required T el}) {
    final node = new Node(el: el);

    if (_head == null) {
      _head = node;
    } else {
      _tail?.next = node;
    }

    _tail = node;

    _counter++;
    return length;
  }

  Node<T>? _getNode({required int index}) {
    final node = getNode<T>(listHead: _head, listLength: _counter, index: index);
    return node;
  }

  int? insert({required T el, int index = 0}) {
    if (index >= 0 && index <= _counter) {
      final node = new Node(el: el);

      if (index == 0) {
        if (_head == null) {
          _head = node;
          _tail = node;
        } else {
          node.next = _head;
          _head = node;
        }
      } else if (index == length) {
        _tail!.next = node; //* new node
        _tail = node; //* update new node pointer

      } else {
        final ancestor = _getNode(index: index - 1)!;
        final current = ancestor.next;
        node.next = current;
        ancestor.next = node;
      }

      _counter++;
      return length;
    }
  }

  @override
  String toString() {
    var pointer = _head;

    if (pointer == null) {
      return '';
    }

    String str = pointer.el.toString();
    pointer = pointer.next;

    for (var i = 0; i <= length && pointer != null; i++) {
      str = str + ' -> ' + pointer.el.toString();
      pointer = pointer.next;
    }
    return str;
  }
}

File: TypeScript/src/data-structures/lists/entities/SinglyLinkedList.ts
import { List } from '../lib/List'

import { Node } from '~/data-structures/models/Node'

import { FilterCallback, MapCallback } from '~/util/types'

import { concatLists, filterList, mapList } from '~/util/functions'

class SinglyLinkedList<T> extends List<T> {
  push(el: T) {
    const node = new Node(el)

    if (!this._head) {
      this._head = node
    } else {
      this._tail.next = node
    }

    this._tail = node

    this._length++
    return this._length
  }

  insert(el: T, index = 0) {
    if (index >= 0 && index <= this._length) {
      const node = new Node(el)

      if (!index) {
        if (!this._head) {
          this._head = node
          this._tail = node
        } else {
          node.next = this._head
          this._head = node
        }
      } else if (index === this._length) {
        this._tail.next = node
        this._tail = node
      } else {
        const ancestor = this.getNode(index - 1)
        const current = ancestor.next
        node.next = current
        ancestor.next = node
      }

      this._length++
      return this._length
    }
  }

  removeAt(index: number) {
    if (this._head && index >= 0 && index < this._length) {
      let pointer = this._head

      if (!index) {
        this._head = this._head.next

        if (!this._head) {
          this._tail = null
        }
      } else {
        const ancestor = this.getNode(index - 1)
        pointer = ancestor.next
        ancestor.next = pointer.next
        this._tail = ancestor
      }

      this._length--
      return pointer.el
    }
  }

  toString() {
    let pointer = this._head

    if (!pointer) {
      return ''
    }

    let str = `${pointer.el}`
    pointer = pointer.next

    for (let i = 1; i < this._length && pointer; i++) {
      str = `${str} -> ${pointer.el}`
      pointer = pointer.next
    }

    return `${str} ->`
  }

  concat(list: List<T>) {
    const newList = new SinglyLinkedList<T>()
    concatLists(this, list, newList)

    return newList
  }

  filter(callbackFn: FilterCallback<T>) {
    const newList = new SinglyLinkedList<T>()
    filterList(this, newList, callbackFn)

    return newList
  }

  map(callbackFn: MapCallback<T>) {
    const newList = new SinglyLinkedList<T>()
    mapList(this, newList, callbackFn)

    return newList
  }
}

export { SinglyLinkedList as LinkedList }
